web/vibe/web/common.d

```diff
@@ -685,33 +685,180 @@ package enum ParamResult {
+			{
+				if(k.length < minLength)
+					// sanity check to prevent out of bounds
+					continue;
+				if(k.startsWith(fieldname) && k[fieldname.length] == indexSep)
+				{
+					// found a potential match
+					string key = k[fieldname.length + 1 .. $];
+					size_t idx;
+					if(key == indexTrailer)
+					{
+						static if(isSimpleElement)
+						{
+							// this is a non-indexed array item. Find an empty slot, or expand the array
+							import std.algorithm : countUntil;
+							idx = seen[].countUntil(false);
+							static if(isStaticArray!T)
+							{
+								if(idx == size_t.max)
+								{
+									// ignore extras, and we know there are no more matches to come.
+									break;
+								}
+							}
+							else if(idx == size_t.max)
+							{
+								// append to the end.
+								idx = dst.length;
+							}
+						}
+						else
+						{
+							// not valid for non-simple elements.
+							continue;
+						}
+					}
+					else
+					{
+						import std.conv;
+						idx = key.parse!size_t;
+						static if(isStaticArray!T)
+						{
+							if(idx > T.length)
+								// keep existing behavior, ignore extras
+								continue;
+						}
+						else if(idx > MAX_ARR_INDEX)
+						{
+							// Getting a bit large, we don't want to allow DOS attacks.
+							err.field = k;
+							err.text = "Maximum index exceeded";
+							return ParamResult.error;
+						}
+						static if(isSimpleElement)
+						{
+							if(key != indexTrailer)
+								// this can't be a match, simple elements are parsed from
+								// the string, there should be no more to the key.
+								continue;
+						}
+						else
+						{
+							// ensure there's more than just the index trailer
+							if(key.length == indexTrailer.length || !key.startsWith(indexTrailer))
+								// not a valid entry. ignore this entry to preserve existing behavior.
+								continue;
+						}
+					}
+
+					static if(!isStaticArray!T)
+					{
+						// check to see if we need to expand the array
+						if(dst.length <= idx)
+						{
+							dst.length = idx + 1;
+							seen.length = idx + 1;
+						}
+					}
+
+					if(seen[idx])
+					{
+						// don't store it twice
+						continue;
+					}
+					seen[idx] = true;
+
+					static if(isSimpleElement)
+					{
+						auto result = processFormParam(v, k, dst[idx], err);
+					}
+					else
+					{
+						auto subFieldname = k[0 .. $ - key.length + indexTrailer.length];
+						auto result = readFormParamRec(req, dst[idx], subFieldname, true, style, err);
+					}
+					if(result != ParamResult.ok)
+						return result;
+				}
+			}
+
+			return ParamResult.ok;
+		}
+
+		if(processItems(req.form) == ParamResult.error)
+			return ParamResult.error;
+		if(processItems(req.query) == ParamResult.error)
+			return ParamResult.error;
+
+		// make sure all static array items have been seen
+		static if(isStaticArray!T)
+		{
+			import std.algorithm : canFind;
+			if(seen[].canFind(false))
+				return ParamResult.error;
```

Comment:
Looking at this I realize this is a different behavior. The previous behavior asserted on this, instead of returning an error. Which means non-release code would throw an Error, whereas release code would let it slide. Not sure of the thoughts on this.

Reply:
Do you mean line 712 in the original code? That assertion should always be true, because `readFormParamRect` is called with `required==true` and thus should never return `ParamResult.skipped`. In that case, the new behavior would be consistent.

Reply:
Yes, you are right! But here is something I forgot. I didn't fill in the error struct as before. I can fix, but I assume this means CI will restart (and it's all green right now).

Maybe better on a quick fixup?

Reply:
I fixed it, good to merge. May want to wait for the CI to finish.